File: frontend/src/api/swap.ts
import request from '@/utils/request'

export interface SwapRequestItem {
  id: number
  request_no: string
  swap_type: string
  requester_id: number
  requester_name: string | null
  requester_schedule_id: number
  requester_schedule_date: string | null
  requester_shift_name: string | null
  target_id: number | null
  target_name: string | null
  target_schedule_id: number | null
  target_schedule_date: string | null
  target_shift_name: string | null
  claimer_id: number | null
  claimer_name: string | null
  reason: string | null
  status: string
  approved_by: number | null
  approver_name: string | null
  approved_at: string | null
  approve_comment: string | null
  created_at: string | null
  updated_at: string | null
}

export interface SwapCreateParams {
  swap_type: string
  requester_schedule_id: number
  target_id?: number
  target_schedule_id?: number
  reason?: string
}

export function getSwapList(params: {
  role?: string
  status?: string
  swap_type?: string
  page?: number
  page_size?: number
}) {
  return request.get('/api/swaps', { params })
}

export function getAllSwapList(params: {
  status?: string
  swap_type?: string
  page?: number
  page_size?: number
}) {
  return request.get('/api/swaps/all', { params })
}

export function getSwapDetail(id: number) {
  return request.get(`/api/swaps/${id}`)
}

export function createSwap(data: SwapCreateParams) {
  return request.post('/api/swaps', data)
}

export function confirmSwap(id: number) {
  return request.put(`/api/swaps/${id}/confirm`)
}

export function claimSwap(id: number) {
  return request.put(`/api/swaps/${id}/claim`)
}

export function approveSwap(id: number, comment?: string) {
  return request.put(`/api/swaps/${id}/approve`, { approve_comment: comment })
}

export function rejectSwap(id: number, comment?: string) {
  return request.put(`/api/swaps/${id}/reject`, { approve_comment: comment })
}

export function cancelSwap(id: number) {
  return request.put(`/api/swaps/${id}/cancel`)
}

export function getMySchedules(params: { start_date?: string; end_date?: string }) {
  return request.get('/api/schedules', { params })
}

File: backend/app/schemas/swap.py
"""调班管理 Schema"""

from __future__ import annotations

from datetime import date as _date, datetime as _datetime
from typing import Optional
from pydantic import BaseModel, Field


# ==================== 请求 Schema ====================

class SwapRequestCreate(BaseModel):
    """发起调班申请"""
    swap_type: str = Field(..., description="调班类型：specified（指定换班）/ open（开放换班）")
    requester_schedule_id: int = Field(..., gt=0, description="发起人的排班记录ID")
    target_id: Optional[int] = Field(None, description="被换人用户ID（指定换班时必填）")
    target_schedule_id: Optional[int] = Field(None, description="被换人的排班记录ID（指定换班时必填）")
    reason: Optional[str] = Field(None, max_length=500, description="申请原因")


class SwapApproveRequest(BaseModel):
    """审批调班申请"""
    approve_comment: Optional[str] = Field(None, max_length=500, description="审批意见")


# ==================== 响应 Schema ====================

class SwapRequestResponse(BaseModel):
    """调班申请响应"""
    id: int
    request_no: str
    swap_type: str
    requester_id: int
    requester_name: Optional[str] = None
    requester_schedule_id: int
    requester_schedule_date: Optional[_date] = None
    requester_shift_name: Optional[str] = None
    target_id: Optional[int] = None
    target_name: Optional[str] = None
    target_schedule_id: Optional[int] = None
    target_schedule_date: Optional[_date] = None
    target_shift_name: Optional[str] = None
    claimer_id: Optional[int] = None
    claimer_name: Optional[str] = None
    reason: Optional[str] = None
    status: str
    approved_by: Optional[int] = None
    approver_name: Optional[str] = None
    approved_at: Optional[_datetime] = None
    approve_comment: Optional[str] = None
    created_at: Optional[_datetime] = None
    updated_at: Optional[_datetime] = None

    class Config:
        from_attributes = True


class SwapListResponse(BaseModel):
    """调班列表响应"""
    items: list[SwapRequestResponse]
    total: int
